clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerMetadata.java

```diff
@@ -68,7 +68,7 @@ public boolean allowAutoTopicCreation() {
 
     @Override
     public synchronized MetadataRequest.Builder newMetadataRequestBuilder() {
-        if (subscription.hasPatternSubscription())
+        if (subscription.hasPatternSubscription() || subscription.hasRe2JPatternSubscription())
```

Comment:
Why do we need to request all topics here? When a new regex is used, it is not different from subscribing to topic names from the consumer perspective as the consumer just receives some newly assigned topic id-partitions in both cases.

Reply:
when the member gets the assignment (with topic Ids only), it requests a metadata update. If not requesting all topics, that metadata request will only include the topic names from the subscription (empty at this point).

https://github.com/apache/kafka/blob/e1ba01d21440a1de011c116a0bb6933656200009/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerMetadata.java#L74

From the client point of view, we only have a pattern, and a list of topic IDs received in assignment, so requesting metadata for all topics. Makes sense?

Reply:
Ouch. It is not supposed to work like this. One of the main benefit of moving the regex to the server side is actually to avoid those "request all metadata" requests. We should be able to only request the metadata of the assigned topic ids regardless of whether the consumer subscribed via names or regex. Of course, the local regex resolution requires this. I understand that it is a limitation of the current implementation of the subscription because it seems that we request based on the subscribed topic names in the subscription state. I suppose that it is not possible to change this in this PR. However, we should file a jira (if not already done) to refactor the subscription state for the new consumer. What do you think?

Reply:
uhm makes total sense, I missed that. I will create and take the jira. The gap is mainly in the subscription state I believe, where we don't keep topic IDs (that unresolved assignment is kept only in the MembershipMgr). Then we also need the ConsumerMetadata to build the metadata request based on the topics Ids if we're using the new pattern (the request builder based on topics IDs already exists but only used in tests).

Reply:
> One of the main benefit of moving the regex to the server side is actually to avoid those "request all metadata" requests.

Let's try to imagine the solution. When users are using server-side regex, `ConsumerMetadata` needs to send a `MetadataRequest` to request only node information (for the coordinator). Then, `ConsumerHeartbeatRequestManager` uses the heartbeat to obtain assignments from the coordinator. Finally, `ConsumerMetadata` can generate a `MetadataRequest` with the correct topic IDs. Is this correct?

Reply:
This is correct for when the consumer subscribed by topic names or by server-side regex. They are actually the same. The only exception is the client-side regex but this will be deprecated soon.

Overall, I would say that we should update the subscription state to operate with topic ids. This could also allow use to simplify the membership manager. We have been talking about doing this for a while now but we never this. It may be the time.

Reply:
> The only exception is the client-side regex but this will be deprecated soon.

Is there already a Jira ticket for this? If not, I’d like to create one to deprecate the `subscribe(Pattern)` and `enforceRebalance` in accordance with KIP-848

Reply:
I need to create all the deprecation tickets. I was holding them off because we should not do them now. I target 4.1 to deprecate them.

Reply:
> I target 4.1 to deprecate them.

Out of curiosity, why aren’t they being deprecated if we aim to set the async consumer as the default in version 4.0? Specifically, `Consumer#enforceRebalance` is essentially a no-op in `AsyncConsumer`